File: src/main/java/algo/backtrack/LetterCombinationsOfAPhoneNumber_17.java
package algo.backtrack;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class LetterCombinationsOfAPhoneNumber_17 {
    private static final Map<Character, String> letterMap = new HashMap<>() {
        {
            put('2', "abc");
            put('3', "def");
            put('4', "ghi");
            put('5', "jkl");
            put('6', "mno");
            put('7', "pqrs");
            put('8', "tuv");
            put('9', "wxyz");
        }
    };

    public static List<String> letterCombinations(String digits) {
        var combinations = new ArrayList<String>();
        if (digits.isEmpty()) {
            return combinations;
        }
        backtrack(digits, "", combinations);
        return combinations;
    }

    private static void backtrack(String sub, String path, List<String> combinations) {
        var digit = sub.charAt(0);
        var letters = letterMap.get(digit);
        for (var j = 0; j < letters.length(); j++) {
            path += letters.charAt(j);
            if (sub.length() == 1) {
                combinations.add(path);
            } else {
                backtrack(sub.substring(1), path, combinations);
            }
            path = path.substring(0, path.length() - 1);
        }
    }

    public static void main(String[] args) {
        System.out.println(letterCombinations("2"));
        System.out.println(letterCombinations("23"));
    }
}

File: src/main/kotlin/algo/backtrack/LetterCombinationsOfAPhoneNumber_17.kt
package algo.backtrack

class KtLetterCombinationsOfAPhoneNumber_17 {
    fun letterCombinations(digits: String): List<String> {
        val all: MutableList<String> = mutableListOf()
        val letterMap: Map<Char, CharArray> = mapOf(
            '2' to charArrayOf('a', 'b', 'c'),
            '3' to charArrayOf('d', 'e', 'f'),
            '4' to charArrayOf('g', 'h', 'i'),
            '5' to charArrayOf('j', 'k', 'l'),
            '6' to charArrayOf('m', 'n', 'o'),
            '7' to charArrayOf('p', 'q', 'r', 's'),
            '8' to charArrayOf('t', 'u', 'v'),
            '9' to charArrayOf('w', 'x', 'y', 'z')
        )
        fun backtrack(digits: String, combination: String) {
            if (digits == "") {
                if (combination != "") {
                    all.add(combination)
                }
                return
            }
            val digit = digits[0]
            letterMap[digit]!!.forEach {
                backtrack(digits.substring(1), "$combination$it")
            }
        }
        backtrack(digits, "")
        return all
    }
}

fun main() {
    val letterCombination = KtLetterCombinationsOfAPhoneNumber_17()
    println(letterCombination.letterCombinations("23"))
    println(letterCombination.letterCombinations(""))
}